File: python/demos/lagrange_tet_demo.py
# from docs/Source_Issue_43.py

import numpy as np
import vtk

# Making a 6-th order tetrahedron
order = 6
n_points = int((order + 1) * (order + 2) * (order + 3) / 6)

# Create a tetrahedron with given number of points.
# Internally, Lagrange cells compute their order according to
# the number of points they hold
tet = vtk.vtkLagrangeTetra()
tet.GetPointIds().SetNumberOfIds(n_points)
tet.GetPoints().SetNumberOfPoints(n_points)
tet.Initialize()

point = [0., 0., 0.]
barycentric_index = [0, 0, 0, 0]

for i in range(n_points):
    tet.GetPointIds().SetId(i, i)
    tet.ToBarycentricIndex(i, barycentric_index)
    for j in range(3):
        point[j] = barycentric_index[j] / order
    tet.GetPoints().SetPoint(i, point[0], point[1], point[2])

tets = vtk.vtkCellArray()
tets.InsertNextCell(tet)

# TODO for Ron:  Compare this to the other ugrid examples.
#  Pretty much boilerplate from here on down
ugrid = vtk.vtkUnstructuredGrid()
ugrid.SetPoints(tet.GetPoints())
ugrid.InsertNextCell(tet.GetCellType(), tet.GetPointIds())

colors = vtk.vtkNamedColors()

mapper = vtk.vtkDataSetMapper()
mapper.SetInputData(ugrid)

actor = vtk.vtkActor()
actor.SetMapper(mapper)
actor.GetProperty().SetColor(colors.GetColor3d("Peacock"))
actor.GetProperty().EdgeVisibilityOn()

ren = vtk.vtkRenderer()
ren_win = vtk.vtkRenderWindow()
ren_win.AddRenderer(ren)
iren = vtk.vtkRenderWindowInteractor()
iren.SetRenderWindow(ren_win)

ren.AddActor(actor)
ren.SetBackground(colors.GetColor3d("Beige"))

iren.Initialize()
ren_win.Render()
iren.Start()

File: cxx/demos/lagrange_tet_demo.cpp
#include "vtkActor.h"
#include "vtkCellArray.h"
#include "vtkDataSetMapper.h"
#include "vtkLagrangeTetra.h"
#include "vtkNamedColors.h"
#include "vtkPoints.h"
#include "vtkProperty.h"
#include "vtkRenderWindow.h"
#include "vtkRenderWindowInteractor.h"
#include "vtkRenderer.h"
#include "vtkSmartPointer.h"
#include "vtkUnstructuredGrid.h"

int main()
{
  int order = 6;
  int nPoints = int((order + 1) * (order + 2) * (order + 3) / 6);

  vtkNew<vtkLagrangeTetra> tet;
  tet->GetPointIds()->SetNumberOfIds(nPoints);
  tet->GetPoints()->SetNumberOfPoints(nPoints);
  tet->Initialize();

  double pt[] = { 0.0, 0.0, 0.0 };
  long long barycentricIndex[] = { 0, 0, 0, 0 };

  for (int i = 0; i < nPoints; ++i)
  {
    tet->GetPointIds()->SetId(i, i);
    tet->ToBarycentricIndex(i, barycentricIndex);
    for (int j = 0; j < 3; ++j)
    {
      pt[j] = double(barycentricIndex[j]) / order;
    }
    tet->GetPoints()->SetPoint(i, pt);
  }

  vtkNew<vtkCellArray> tets;
  tets->InsertNextCell(tet);

  vtkNew<vtkUnstructuredGrid> ugrid;
  ugrid->SetPoints(tet->GetPoints());
  ugrid->InsertNextCell(tet->GetCellType(), tet->GetPointIds());

  vtkNew<vtkDataSetMapper> mapper;
  mapper->SetInputData(ugrid);

  vtkNew<vtkActor> actor;
  actor->SetMapper(mapper);
  actor->GetProperty()->EdgeVisibilityOn();

  vtkNew<vtkRenderer> ren;
  vtkNew<vtkRenderWindow> renWin;
  renWin->AddRenderer(ren);
  vtkNew<vtkRenderWindowInteractor> iren;
  iren->SetRenderWindow(renWin);

  ren->AddActor(actor);
  ren->SetBackground(249, 228, 183); // beige

  renWin->Render();

  iren->Initialize();
  iren->Start();
}